intel/multus-cni
# Adding conflist support in Multus

Referred to #29 issue.

**How to test the PR ?**

1. Check the `07_ptp_tuning_conflist.yml` in the PR, this conflist have [ptp ](https://github.com/containernetworking/plugins/tree/master/plugins/main/ptp)and [tuning](https://github.com/containernetworking/plugins/tree/master/plugins/meta/tuning). The ptp is to provide the point-to-point network connection and tuning plugin can change some system controls (sysctls) in the network namespace.

```
# cat examples/npwg-demo-1/07_ptp_tuning_conflist.yml 
---                                                                                    
apiVersion: "k8s.cni.cncf.io/v1"                                                       
kind: NetworkAttachmentDefinition                                                      
metadata:                                                                              
  name: ptp-tuning-conflist                                                            
spec:                                                                                  
  config: '{                                                                           
        "cniVersion": "0.3.1",                                                         
        "name": "ptp-tuning-conflist",                                                 
        "plugins": [{                                                                  
                        "type": "ptp",                                                 
                        "ipMasq": true,                                                
                        "mtu": 512,                                                    
                        "ipam": {                                                      
                                "type": "host-local",                                  
                                "subnet": "172.16.0.0/24"                              
                        },                                                             
                        "dns": {                                                       
                                "nameservers": ["172.16.1.1"]                          
                        }                                                              
                },                                                                     
                {                                                                      
                        "name": "mytuning",                                            
                        "type": "tuning",                                              
                        "sysctl": {                                                    
                                "net.core.somaxconn": "500"                            
                        }                                                              
                }                                                                      
        ]                                                                              
}'
```
2. Create the conflist network object
```
# kubectl create -f examples/npwg-demo-1/07_ptp_tuning_conflist.yml
networkattachmentdefinition.k8s.cni.cncf.io/ptp-tuning-conflist created 
```
3. Pod have ptp tuning conflist network object
```                                                                               
# cat examples/npwg-demo-1/15_pod_case5.yml
---
apiVersion: v1
kind: Pod
metadata:
  name: pod-case-05
  annotations:
    k8s.v1.cni.cncf.io/networks: '[
            { "name": "ptp-tuning-conflist" }
    ]'
spec:
  containers:
  - name: pod-case-05
    image: docker.io/centos/tools:latest
    command:
    - /sbin/init
```
```	
# kubectl create -f examples/npwg-demo-1/15_pod_case5.yml
pod/pod-case-05 created
```
```
# kubectl get pods
NAME          READY     STATUS    RESTARTS   AGE
pod-case-05   1/1       Running   0          1m
```
4. Pod having ptp as the 2nd network
```
# kubectl exec -it pod-case-05 -- ifconfig
eth0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1450
        inet 192.168.49.121  netmask 255.255.255.0  broadcast 0.0.0.0
        ether 5a:cc:37:37:6b:f8  txqueuelen 0  (Ethernet)
        RX packets 0  bytes 0 (0.0 B)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 1  bytes 42 (42.0 B)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

lo: flags=73<UP,LOOPBACK,RUNNING>  mtu 65536
        inet 127.0.0.1  netmask 255.0.0.0
        loop  txqueuelen 1  (Local Loopback)
        RX packets 0  bytes 0 (0.0 B)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 0  bytes 0 (0.0 B)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

net1: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 512
        inet 172.16.0.4  netmask 255.255.255.0  broadcast 0.0.0.0
        ether 52:1a:cd:65:1f:2a  txqueuelen 0  (Ethernet)
        RX packets 0  bytes 0 (0.0 B)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 0  bytes 0 (0.0 B)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0
```
5. Tuning plugin will set /proc/sys/net/core/somaxconn to 500
```
# kubectl exec -it pod-case-05 -- cat /proc/sys/net/core/somaxconn
500
```
